### PR TITLE
Psl 165 convert uint256 test arith uint256 test

### PR DESCRIPTION
**Pull request by tanlemin91**
- convert uint256_tests.cpp, arith_uint256_tests.cpp to pastel_gtest
- add uint256_tests.cpp, arith_uint256_tests.cpp into pastel_gtest
- remove uint256_tests.cpp, arith_uint256_tests.cpp from test_bitcoin

### CODE REVIEW

**Review by tanlemin91 — COMMENTED** [review]
Please review @a-ok123 @akobrin2

File: src/gtest/test_arith_uint256.cpp [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+#include <stdint.h>
+#include <sstream>
+#include <iomanip>
+#include <limits>
+#include <cmath>
+#include "uint256.h"
+#include "arith_uint256.h"
+#include <string>
+#include "version.h"
+#include "test/test_bitcoin.h"
```
Comment (akobrin1) [inline src/gtest/test_arith_uint256.cpp:16]: remove test_bitcoin.h, the goal is to remove all connections to boost

File: src/gtest/test_arith_uint256.cpp [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+TEST(arith_uint256_tests, basic)
+{
+    EXPECT_EQ(1 , 0+1);
+    // constructor arith_uint256(vector<char>):
+    EXPECT_EQ(R1L.ToString() , ArrToStr(R1Array,32));
+    EXPECT_EQ(R2L.ToString() , ArrToStr(R2Array,32));
+    EXPECT_EQ(ZeroL.ToString() , ArrToStr(ZeroArray,32));
+    EXPECT_EQ(OneL.ToString() , ArrToStr(OneArray,32));
+    EXPECT_EQ(MaxL.ToString() , ArrToStr(MaxArray,32));
+    EXPECT_TRUE(OneL.ToString() != ArrToStr(ZeroArray,32));
```
Comment (akobrin1) [inline src/gtest/test_arith_uint256.cpp:78]: replace EXPECT_TRUE(a != b) with EXPECT_NE(a, b)
and all similar cases below

File: src/gtest/test_arith_uint256.cpp [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+
+    // == and !=
+    EXPECT_TRUE(R1L != R2L);
+    EXPECT_TRUE(ZeroL != OneL);
+    EXPECT_TRUE(OneL != ZeroL);
+    EXPECT_TRUE(MaxL != ZeroL);
+    EXPECT_TRUE(~MaxL == ZeroL);
```
Comment (akobrin1) [inline src/gtest/test_arith_uint256.cpp:85]: this one is EXPECT_EQ

File: src/gtest/test_arith_uint256.cpp [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+    EXPECT_TRUE( ((R1L ^ R2L) ^ R1L) == R2L);
+
+    uint64_t Tmp64 = 0xc4dab720d9c7acaaULL;
+    for (unsigned int i = 0; i < 256; ++i)
+    {
+        EXPECT_TRUE(ZeroL != (OneL << i));
```
Comment (akobrin1) [inline src/gtest/test_arith_uint256.cpp:91]: EXPECT_NE - all 4

File: src/gtest/test_arith_uint256.cpp [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+        EXPECT_TRUE((OneL << i) != ZeroL);
+        EXPECT_TRUE(R1L != (R1L ^ (OneL << i)));
+        EXPECT_TRUE(((arith_uint256(Tmp64) ^ (OneL << i) ) != Tmp64 ));
+    }
+    EXPECT_TRUE(ZeroL == (OneL << 256));
```
Comment (akobrin1) [inline src/gtest/test_arith_uint256.cpp:96]: EXPECT_EQ all below

File: src/gtest/test_arith_uint256.cpp [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+}
+
+TEST(arith_uint256_tests, shifts) // "<<"  ">>"  "<<="  ">>="
+{
+    unsigned char TmpArray[32];
+    arith_uint256 TmpL;
+    for (unsigned int i = 0; i < 256; ++i)
+    {
+        shiftArrayLeft(TmpArray, OneArray, 32, i);
+        EXPECT_TRUE(arith_uint256V(std::vector<unsigned char>(TmpArray,TmpArray+32)) == (OneL << i));
```
Comment (akobrin1) [inline src/gtest/test_arith_uint256.cpp:166]: add #include <vector_types.h> and replace all vector<unsigned char> with v_uint8.
all std:: can be removed in this file - you have "using namespace std" defined

File: src/gtest/test_arith_uint256.cpp [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+TEST(arith_uint256_tests, unaryOperators) // !    ~    -
+{
+    EXPECT_TRUE(!ZeroL);
+    EXPECT_TRUE(!(!OneL));
+    for (unsigned int i = 0; i < 256; ++i)
+        EXPECT_TRUE(!(!(OneL<<i)));
+    EXPECT_TRUE(!(!R1L));
+    EXPECT_TRUE(!(!MaxL));
+
+    EXPECT_TRUE(~ZeroL == MaxL);
```
Comment (akobrin1) [inline src/gtest/test_arith_uint256.cpp:212]: EXPECT_TRUE=> EXPECT_EQ

File: src/gtest/test_arith_uint256.cpp [hunk 8]
```diff
@@ -0,0 +1,567 @@
...
+}
+
+// Check if doing _A_ _OP_ _B_ results in the same as applying _OP_ onto each
+// element of Aarray and Barray, and then converting the result into a arith_uint256.
+#define CHECKBITWISEOPERATOR(_A_,_B_,_OP_)                              \
+    for (unsigned int i = 0; i < 32; ++i) { TmpArray[i] = _A_##Array[i] _OP_ _B_##Array[i]; } \
+    EXPECT_TRUE(arith_uint256V(std::vector<unsigned char>(TmpArray,TmpArray+32)) == (_A_##L _OP_ _B_##L));
+
+#define CHECKASSIGNMENTOPERATOR(_A_,_B_,_OP_)                           \
+    TmpL = _A_##L; TmpL _OP_##= _B_##L; EXPECT_TRUE(TmpL == (_A_##L _OP_ _B_##L));
```
Comment (akobrin1) [inline src/gtest/test_arith_uint256.cpp:231]: this EXPECT_TRUE in macro also to EXPECT_EQ

File: src/gtest/test_arith_uint256.cpp [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+
+TEST(arith_uint256_tests, bignum_SetCompact)
+{
+    arith_uint256 num;
+    bool fNegative;
+    bool fOverflow;
+    num.SetCompact(0, &fNegative, &fOverflow);
+    EXPECT_EQ(num.GetHex(), "0000000000000000000000000000000000000000000000000000000000000000");
+    EXPECT_EQ(num.GetCompact(), 0U);
+    EXPECT_EQ(fNegative, false);
```
Comment (akobrin1) [inline src/gtest/test_arith_uint256.cpp:423]: EXPECT_EQ => EXPECT_FALSE,
all fNegative, fOverflow checks below... some of them EXPECT_TRUE, some EXPECT_FALSE

File: src/gtest/test_arith_uint256.cpp [hunk 10]
```diff
@@ -0,0 +1,567 @@
+// Copyright (c) 2011-2013 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
```
Comment (akobrin1) [inline src/gtest/test_arith_uint256.cpp:6]: you can delete gtest.h here, gmock.h includes gtest.h

File: src/gtest/test_uint256.cpp [hunk 11]
```diff
@@ -0,0 +1,267 @@
+// Copyright (c) 2011-2013 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+#include "arith_uint256.h"
+#include "uint256.h"
+#include "version.h"
+#include "test/test_bitcoin.h"
```
Comment (akobrin1) [inline src/gtest/test_uint256.cpp:7]: remove test_bitcoin.h & gtest.h

**Review by akobrin1 — CHANGES_REQUESTED** [review]
please review all EXPECT_xxxx macros

**Review by akobrin1 — APPROVED** [review]
You could convert some tests to parameterized.
Especially tests with lots of repetitions, like bignum_SetCompact.
See for example test_bech32.cpp. In this test, you could use tuple<uint32_t, string, uint32_t, bool, bool> - simply all parameters used in the test. Then just write the actual test once and define all parameters via INSTANTIATE_TEST_SUITE_P && Values(make_tuple(...), )

**Review by a-ok123 — APPROVED** [review]

**Pull request merged**